OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorLogic.cs

```diff
@@ -0,0 +1,155 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	public class MapEditorLogic
+	{
+		readonly Ruleset rules;
+
+		CPos cachedMouseCursorCell;
+		Rectangle draggedTemplateBounds;
+		ResourceTypeInfo draggedResourceTypeInfo;
+		string draggedActor;
+		int cachedFacing;
+		string cachedOwner;
+
+		[ObjectCreator.UseCtor]
+		public MapEditorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+		{
+			rules = world.Map.Rules;
+
+			var mapEditorWidget = widget.Get<MapEditorWidget>("MAP_EDITOR");
+
+			var dragTilePreview = widget.Get<TerrainTemplatePreviewWidget>("DRAG_TILE_PREVIEW");
+			dragTilePreview.GetScale = () => worldRenderer.Viewport.Zoom;
+			dragTilePreview.IsVisible = () => mapEditorWidget.CurrentPreview == MapEditorWidget.DragType.Tiles;
+
+			var dragLayerPreview = widget.Get<SpriteWidget>("DRAG_LAYER_PREVIEW");
+			dragLayerPreview.GetScale = () => worldRenderer.Viewport.Zoom;
+			dragLayerPreview.IsVisible = () => mapEditorWidget.CurrentPreview == MapEditorWidget.DragType.Layers;
+			dragLayerPreview.GetPalette = () => mapEditorWidget.SelectedResourceTypeInfo.Palette;
+			dragLayerPreview.GetSprite = () =>
+			{
+				var resource = mapEditorWidget.SelectedResourceTypeInfo;
+				var variant = resource.Variants.FirstOrDefault();
+				var sequenceProvier = rules.Sequences[world.TileSet.Id];
+				var sequence = sequenceProvier.GetSequence("resources", variant);
+				var frame = sequence.Frames != null ? sequence.Frames.Last() : resource.MaxDensity - 1;
+				return sequence.GetSprite(frame);
+			};
+
+			var dragActorPreview = widget.Get<ActorPreviewWidget>("DRAG_ACTOR_PREVIEW");
+			dragActorPreview.GetScale = () => worldRenderer.Viewport.Zoom;
+			dragActorPreview.IsVisible = () => mapEditorWidget.CurrentPreview == MapEditorWidget.DragType.Actors;
+
+			var ticker = widget.Get<LogicTickerWidget>("EDITOR_TICKER");
+			ticker.OnTick = () =>
+			{
```

Comment:
These hardcoded brush types will need to be replaced with a proper abstraction before we can continue with the next set of editor features (height manipulation, copy/paste, etc), but i've given up on the idea of having this fixed before the initial merge.